service/history/api/startworkflow/api.go

```diff
@@ -41,167 +46,507 @@ import (
+	workflowEventBatches []*persistence.WorkflowEvents
+}
+
+// mutableStateInfo is a container for the relevant mutable state information to generate a start response with an eager
+// workflow task.
+type mutableStateInfo struct {
+	branchToken      []byte
+	lastEventID      int64
+	workflowTaskInfo *workflow.WorkflowTaskInfo
+	hasInflight      bool
+}
+
+// NewStarter creates a new starter, fails if getting the active namespace fails.
+func NewStarter(
+	shardCtx shard.Context,
 	workflowConsistencyChecker api.WorkflowConsistencyChecker,
-) (resp *historyservice.StartWorkflowExecutionResponse, retError error) {
-	namespaceEntry, err := api.GetActiveNamespace(shard, namespace.ID(startRequest.GetNamespaceId()))
+	tokenSerializer common.TaskTokenSerializer,
+	request *historyservice.StartWorkflowExecutionRequest,
+) (*Starter, error) {
+	namespaceEntry, err := api.GetActiveNamespace(shardCtx, namespace.ID(request.GetNamespaceId()))
 	if err != nil {
 		return nil, err
 	}
-	namespaceID := namespaceEntry.ID()
+	return &Starter{
+		shardCtx:                   shardCtx,
+		workflowConsistencyChecker: workflowConsistencyChecker,
+		tokenSerializer:            tokenSerializer,
+		request:                    request,
+		namespace:                  namespaceEntry,
+	}, nil
+}
+
+// prepare applies request overrides, validates the request, and records eager execution metrics.
+func (s *Starter) prepare(ctx context.Context) error {
+	request := s.request.StartRequest
+	metricsHandler := s.shardCtx.GetMetricsHandler()
+
+	api.OverrideStartWorkflowExecutionRequest(request, metrics.HistoryStartWorkflowExecutionScope, s.shardCtx, metricsHandler)
+	err := api.ValidateStartWorkflowExecutionRequest(ctx, request, s.shardCtx, s.namespace, "StartWorkflowExecution")
+	if err != nil {
+		return err
+	}
+
+	if request.RequestEagerExecution {
+		metricsHandler.Counter(metrics.WorkflowEagerExecutionCounter.GetMetricName()).Record(
+			1,
+			metrics.NamespaceTag(s.namespace.Name().String()),
+			metrics.TaskQueueTag(request.TaskQueue.Name),
+			metrics.WorkflowTypeTag(request.WorkflowType.Name),
+		)
+	}
+
+	// Override to false to avoid having to look up the dynamic config throughout the diffrent code paths.
+	if !s.shardCtx.GetConfig().EnableEagerWorkflowStart(s.namespace.Name().String()) {
+		s.recordEagerDenied(eagerStartDeniedReasonDynamicConfigDisabled)
+		request.RequestEagerExecution = false
+	}
+	if s.request.FirstWorkflowTaskBackoff != nil && *s.request.FirstWorkflowTaskBackoff > 0 {
+		s.recordEagerDenied(eagerStartDeniedReasonFirstWorkflowTaskBackoff)
+		request.RequestEagerExecution = false
+	}
+	return nil
+}
+
+func (s *Starter) recordEagerDenied(reason eagerStartDeniedReason) {
+	metricsHandler := s.shardCtx.GetMetricsHandler()
+	metricsHandler.Counter(metrics.WorkflowEagerExecutionDeniedCounter.GetMetricName()).Record(
+		1,
+		metrics.NamespaceTag(s.namespace.Name().String()),
+		metrics.TaskQueueTag(s.request.StartRequest.TaskQueue.Name),
```

Comment:
This is another place where using a parsed type instead of the raw StartRequest is better because it prevents Law of Demeter violations here and elsewhere